numpy/_core/src/multiarray/shape.c

```diff
@@ -141,8 +141,9 @@ PyArray_Resize(PyArrayObject *self, PyArray_Dims *newshape, int refcheck,
     }
 
     if (newnbytes > oldnbytes && PyArray_ISWRITEABLE(self)) {
-        /* Fill new memory with zeros */
-        if (PyDataType_FLAGCHK(PyArray_DESCR(self), NPY_ITEM_REFCOUNT)) {
+        /* Fill new memory with zeros (PyLong zero for object arrays) */
+        if (PyDataType_ISLEGACY(PyArray_DESCR(self)) &&
+            PyDataType_FLAGCHK(PyArray_DESCR(self), NPY_ITEM_REFCOUNT)) {
```

Comment:
Sorry, but I think we have to fix this properly one way or another.  Resize (for worse probably) effectively uses `zeros()` logic when expanding the array.

Zeros logic is implemented now with:
```
        PyArrayMethod_GetTraverseLoop *get_fill_zero_loop =
            NPY_DT_SLOTS(NPY_DTYPE(descr))->get_fill_zero_loop;
        if (get_fill_zero_loop != NULL) {
            if (get_fill_zero_loop(
                    NULL, descr, 1, descr->elsize, &(fill_zero_info.func),
                    &(fill_zero_info.auxdata), &zero_flags) < 0) {
                goto fail;
            }
        }
```
etc., see `PyArray_NewFromDescr_int`.  We have to aim for consolidating these paths or the mess this is will never go away.  In particularly that internal `_putzero` shouldn't exist!
(It must be a duplicate or probably triplicate of some fill zero function!)

Reply:
Thanks! I forgot about this API but you're right it's clearly the correct thing to use.

Reply:
One thing that may be nice would be to create a `PyArray_ZeroFillBuffer()` helper similar to the `PyArray_ClearBuffer()` one.